homeassistant/bootstrap.py

```diff
@@ -436,14 +450,18 @@ def async_from_config_dict(config: Dict[str, Any],
     # Setup the components
     dependency_blacklist = loader.DEPENDENCY_BLACKLIST - set(components)
 
+    tasks = []
+    component_list = []
     for domain in loader.load_order_components(components):
         if domain in dependency_blacklist:
             raise HomeAssistantError(
                 '{} is not allowed to be a dependency'.format(domain))
+        tasks.append(_async_setup_component(hass, domain, config))
```

Comment:
I think that the fact that we keep track of internal tasks and things like `_async_init_setup_data` should be abstracted away. `_async_setup_component` should do the check if it is already in progress and yield from the correct event or whatever we use to signal it.